File: src/main/scala/com/loquor/sort/QuicklySort.scala
package com.loquor.sort

object QuicklySort {


  def main(args: Array[String]): Unit = {
    val arr = Array(2, 4, 3, 6, 7, 2, 9, 5, 23, 32, 12, 555, 123, 32, 765);
    val length = arr.length
    val start = System.currentTimeMillis();
    sort(arr, 0, length - 1)
    println(arr.mkString("Array(", ", ", ")"))
    println(System.currentTimeMillis() - start)
  }

  def sort(arr: Array[Int], start: Int, end: Int): Unit = {
    if (start > end) {
      return
    }
    val temp = arr(start)
    var left = start
    var right = end

    while (left < right) {

      while (left < right && arr(right) > temp) {
        right = right - 1
      }

      if (left < right) {
        arr(left) = arr(right);
        left = left + 1
      }

      while (left < right && arr(left) >= temp) {
        left = left + 1
      }

      if (left < right) {
        arr(right) = arr(left)
        right = right - 1
      }
    }
    arr(left) = temp
    sort(arr, start, left - 1)
    sort(arr, left + 1, end)
  }
}

File: src/main/java/com/liquor/sort/QuicklySort.java
package com.liquor.sort;

import java.util.Arrays;

//https://zhuanlan.zhihu.com/p/35946897
public class QuicklySort {

    public static void main(String[] args) {
        int[] arr = new int[]{3, 5, 3, 2, 1};
        final int length = arr.length;
        final long start = System.currentTimeMillis();
        sort(arr, 0, length - 1);
        System.out.println(Arrays.toString(arr));
        System.out.println(System.currentTimeMillis() - start);

    }

    private static void sort(int[] arr, int start, int end) {
        if (start < end) {
            int tmp = arr[start];
            int left = start;
            int right = end;

            while (left < right) {

                while (left < right && arr[right] >= tmp) {
                    right--;
                }

                if (left < right) {
                    arr[left] = arr[right];
                    left++;
                }


                while (left < right && arr[left] < tmp) {
                    left++;
                }

                if (left < right) {
                    arr[right] = arr[left];
                    right--;
                }

            }
            arr[left] = tmp;
            sort(arr, start, left - 1);
            sort(arr, left + 1, end);
        }
    }
}
